**TASK-6**

**ASSIGNMENT-8**

**How Much is Your Car Worth?**

How Much is Your Car Worth?

Data about the retail price of 2005 General Motors cars can be found in car_data.csv.

The columns are:

1)Price: suggested retail price of the used 2005 GM car in excellent condition.

2)Mileage: number of miles the car has been driven

3)Make: manufacturer of the car such as Saturn, Pontiac, and Chevrolet

4)Model: specific models for each car manufacturer such as Ion, Vibe, Cavalier

5)Trim (of car): specific type of car model such as SE Sedan 4D, Quad Coupe 2D
Type: body type such as sedan, coupe, etc.

6)Cylinder: number of cylinders in the engine

7)Liter: a more specific measure of engine size

8)Doors: number of doors

9)Cruise: indicator variable representing whether the car has cruise control (1 = cruise)

10)Sound: indicator variable representing whether the car has upgraded speakers (1 = upgraded)

11)Leather: indicator variable representing whether the car has leather seats (1 = leather)

## Tasks, Part 1

1. Find the linear regression equation for mileage vs price.
2. Chart the original data and the equation on the chart.
3. Find the equation's $R^2$ score (use the `.score` method) to determine whether the
equation is a good fit for this data. (0.8 and greater is considered a strong correlation.)

## Tasks, Part 2

1. Use mileage, cylinders, liters, doors, cruise, sound, and leather to find the linear regression equation.
2. Find the equation's $R^2$ score (use the `.score` method) to determine whether the
equation is a good fit for this data. (0.8 and greater is considered a strong correlation.)
3. Find the combination of the factors that is the best predictor for price.

## Tasks, Hard Mode

1. Research dummy variables in scikit-learn to see how to use the make, model, and body type.
2. Find the best combination of factors to predict price.

In [64]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn import linear_model

In [65]:
data=pd.read_csv("/content/car_data.csv")

In [66]:
data.head()

Price  Mileage   Make    Model      Trim   Type  Cylinder  Liter  \
0  17314.103129     8221  Buick  Century  Sedan 4D  Sedan         6    3.1   
1  17542.036083     9135  Buick  Century  Sedan 4D  Sedan         6    3.1   
2  16218.847862    13196  Buick  Century  Sedan 4D  Sedan         6    3.1   
3  16336.913140    16342  Buick  Century  Sedan 4D  Sedan         6    3.1   
4  16339.170324    19832  Buick  Century  Sedan 4D  Sedan         6    3.1   

   Doors  Cruise  Sound  Leather  
0      4       1      1        1  
1      4       1      1        0  
2      4       1      1        0  
3      4       1      0        0  
4      4       1      0        1

In [67]:
data.corr()

Price   Mileage  Cylinder     Liter     Doors    Cruise  \
Price     1.000000 -0.143051  0.569086  0.558146 -0.138750  0.430851   
Mileage  -0.143051  1.000000 -0.029461 -0.018641 -0.016944  0.025037   
Cylinder  0.569086 -0.029461  1.000000  0.957897  0.002206  0.354285   
Liter     0.558146 -0.018641  0.957897  1.000000 -0.079259  0.377509   
Doors    -0.138750 -0.016944  0.002206 -0.079259  1.000000 -0.047674   
Cruise    0.430851  0.025037  0.354285  0.377509 -0.047674  1.000000   
Sound    -0.124348 -0.026146 -0.089704 -0.065527 -0.062530 -0.091730   
Leather   0.157197  0.001005  0.075520  0.087332 -0.061969 -0.070573   

             Sound   Leather  
Price    -0.124348  0.157197  
Mileage  -0.026146  0.001005  
Cylinder -0.089704  0.075520  
Liter    -0.065527  0.087332  
Doors    -0.062530 -0.061969  
Cruise   -0.091730 -0.070573  
Sound     1.000000  0.165444  
Leather   0.165444  1.000000

In [68]:
data.shape

(804, 12)

In [69]:
data.isnull().sum()

Price       0
Mileage     0
Make        0
Model       0
Trim        0
Type        0
Cylinder    0
Liter       0
Doors       0
Cruise      0
Sound       0
Leather     0
dtype: int64

In [70]:
X=data.iloc[:,1] #X=mileage(independent variable)
y=data.iloc[:,0] #y=price(dependent variable)

In [71]:
X=np.array(X).reshape((-1,1))
y=np.array(y).reshape((-1,1))

In [72]:
X

array([[ 8221],
       [ 9135],
       [13196],
       [16342],
       [19832],
       [22236],
       [22576],
       [22964],
       [24021],
       [27325],
       [10237],
       [15066],
       [16633],
       [19800],
       [23359],
       [23765],
       [24008],
       [26034],
       [32896],
       [34665],
       [ 8970],
       [10577],
       [14363],
       [16508],
       [18908],
       [19467],
       [21394],
       [22596],
       [24469],
       [25245],
       [15755],
       [16235],
       [18147],
       [22102],
       [23612],
       [24439],
       [25212],
       [26779],
       [27096],
       [29649],
       [ 2992],
       [10036],
       [18339],
       [19592],
       [21797],
       [23852],
       [25040],
       [25052],
       [25464],
       [27830],
       [17353],
       [17787],
       [19562],
       [20137],
       [20158],
       [23767],
       [24323],
       [27839],
       [29187],
       [31324],
       [ 2308],
       [10014],
       [

In [73]:
y

array([[17314.1031289 ],
       [17542.03608328],
       [16218.84786194],
       [16336.91314005],
       [16339.17032393],
       [15709.05282108],
       [15230.00338985],
       [15048.04218412],
       [14862.0938696 ],
       [15295.01826688],
       [21335.85248489],
       [20538.08751023],
       [20512.09409124],
       [19924.1590524 ],
       [19774.24906607],
       [19344.16553736],
       [19105.13012442],
       [18543.42704537],
       [17808.19899602],
       [17968.83827826],
       [22358.87767812],
       [23785.92252638],
       [22926.09037126],
       [21895.7587842 ],
       [21273.06171392],
       [21460.01395289],
       [21183.12358742],
       [20406.09950289],
       [21058.13957995],
       [19556.89932029],
       [23447.68656101],
       [23547.23944288],
       [23016.00800868],
       [22230.02750256],
       [22625.07356605],
       [21799.17209689],
       [21341.25720707],
       [21683.03146355],
       [20986.01601748],
       [20902.10379431],


In [74]:
#model
from sklearn.linear_model import LinearRegression
model=LinearRegression()
model.fit(X,y)

LinearRegression()

In [75]:
print(model.coef_)#y=mx+c ->>m

[[-0.1725205]]


In [76]:
print(model.intercept_) #y=mx+c->>c

[24764.55900606]


In [77]:
model.score(X,y)

0.02046344732348926

In [129]:
print("Linear Regression Equation for Mileage vs price is",
      "y=",float(model.coef_),"x +",float(model.intercept_))

Linear Regression Equation for Mileage vs price is y= -0.1725205027912493 x + 24764.559006061685


In [79]:
pred=model.predict(X)
pred
from sklearn.metrics import r2_score
print("The R2 score of mileage column is :",r2_score(y,pred))

The R2 score of mileage column is : 0.02046344732348926


**Task-2**

In [88]:
X1=data.iloc[:,[1,6,7,8,9,10,11]]
y1=data.iloc[:,0]

In [89]:
X1=np.array(X1).reshape((-1,7))
y1=np.array(y1).reshape((-1,1))

In [90]:
X1[0]

array([8.221e+03, 6.000e+00, 3.100e+00, 4.000e+00, 1.000e+00, 1.000e+00,
       1.000e+00])

In [91]:
#model
from sklearn.linear_model import LinearRegression
model1=LinearRegression()
model1.fit(X1,y1)

LinearRegression()

In [92]:
model1.coef_

array([[-1.69747832e-01,  3.79237893e+03, -7.87220732e+02,
        -1.54274585e+03,  6.28899715e+03, -1.99379528e+03,
         3.34936162e+03]])

In [93]:
model1.intercept_

array([6758.7551436])

In [130]:
print("Linear Regression Equation using mileage, cylinders, liters, doors, cruise, sound, and leather is",
      "y=",float(model1.coef_[0][0]),"x1 +",float(model1.coef_[0][1]),"x2 +",float(model1.coef_[0][2]),
      "x3 +",float(model1.coef_[0][3]),"x4 +",float(model1.coef_[0][4]),
      "x5 +",float(model1.coef_[0][5]),"x6 +",float(model1.coef_[0][6]),
      "x7 +",float(model1.intercept_))

Linear Regression Equation using mileage, cylinders, liters, doors, cruise, sound, and leather is y= -0.16974783233303692 x1 + 3792.378930782515 x2 + -787.2207316338321 x3 + -1542.745846382068 x4 + 6288.997149849932 x5 + -1993.795275698427 x6 + 3349.361619613555 x7 + 6758.755143598106


In [95]:
pred1=model1.predict(X1)
pred1
#Task2-2Q->>
from sklearn.metrics import r2_score
print("The R2 score of all column is :",r2_score(y1,pred1))

The R2 score of all column is : 0.4462643536728379


In [99]:
from sklearn.decomposition import PCA
s=[]
for i in range(1,8):
  pca=PCA(n_components=i,random_state=0)
  pca.fit(X1)
  x=pca.transform(X1)

  model2=LinearRegression()
  model2.fit(x,y1)

  pred2=model2.predict(x)
  pred2

  s.append(r2_score(y,pred2))
  print("with",i,"columns:",s[i-1])



with 1 columns: 0.020463447494586617
with 2 columns: 0.3504497977596356
with 3 columns: 0.3646064800564265
with 4 columns: 0.3673863089598691
with 5 columns: 0.37068136274290675
with 6 columns: 0.43879954665861454
with 7 columns: 0.446264353672838


**TASK-3**

In [100]:
print(data['Make'].unique())
print(data['Model'].unique())
print(data['Trim'].unique())
print(data['Type'].unique())

['Buick' 'Cadillac' 'Chevrolet' 'Pontiac' 'SAAB' 'Saturn']
['Century' 'Lacrosse' 'Lesabre' 'Park Avenue' 'CST-V' 'CTS' 'Deville'
 'STS-V6' 'STS-V8' 'XLR-V8' 'AVEO' 'Cavalier' 'Classic' 'Cobalt'
 'Corvette' 'Impala' 'Malibu' 'Monte Carlo' 'Bonneville' 'G6' 'Grand Am'
 'Grand Prix' 'GTO' 'Sunfire' 'Vibe' '9_3' '9_3 HO' '9_5' '9_5 HO'
 '9-2X AWD' 'Ion' 'L Series']
['Sedan 4D' 'CX Sedan 4D' 'CXL Sedan 4D' 'CXS Sedan 4D' 'Custom Sedan 4D'
 'Limited Sedan 4D' 'Special Ed Ultra 4D' 'DHS Sedan 4D' 'DTS Sedan 4D'
 'Hardtop Conv 2D' 'LS Hatchback 4D' 'LS Sedan 4D' 'LT Hatchback 4D'
 'LT Sedan 4D' 'SVM Hatchback 4D' 'SVM Sedan 4D' 'Coupe 2D' 'LS Coupe 2D'
 'LS Sport Coupe 2D' 'LS Sport Sedan 4D' 'Conv 2D' 'SS Sedan 4D'
 'LS MAXX Hback 4D' 'LT MAXX Hback 4D' 'MAXX Hback 4D' 'LT Coupe 2D'
 'SS Coupe 2D' 'GXP Sedan 4D' 'SE Sedan 4D' 'SLE Sedan 4D' 'GT Sedan 4D'
 'GT Coupe 2D' 'GTP Sedan 4D' 'AWD Sportwagon 4D' 'GT Sportwagon'
 'Sportwagon 4D' 'Linear Conv 2D' 'Linear Sedan 4D' 'Aero Conv 2D'
 'Aero 

In [101]:
encoded_data = pd.get_dummies(data, columns = ['Make','Model','Trim','Type'])

In [104]:
encoded_data.head()

Price  Mileage  Cylinder  Liter  Doors  Cruise  Sound  Leather  \
0  17314.103129     8221         6    3.1      4       1      1        1   
1  17542.036083     9135         6    3.1      4       1      1        0   
2  16218.847862    13196         6    3.1      4       1      1        0   
3  16336.913140    16342         6    3.1      4       1      0        0   
4  16339.170324    19832         6    3.1      4       1      0        1   

   Make_Buick  Make_Cadillac  ...  Trim_SVM Hatchback 4D  Trim_SVM Sedan 4D  \
0           1              0  ...                      0                  0   
1           1              0  ...                      0                  0   
2           1              0  ...                      0                  0   
3           1              0  ...                      0                  0   
4           1              0  ...                      0                  0   

   Trim_Sedan 4D  Trim_Special Ed Ultra 4D  Trim_Sportwagon 4D  \
0              1                         0                   0   
1              1                         0                   0   
2              1                         0                   0   
3              1                         0                   0   
4              1                         0                   0   

   Type_Convertible  Type_Coupe  Type_Hatchback  Type_Sedan  Type_Wagon  
0                 0           0               0           1           0  
1                 0           0               0           1           0  
2                 0           0               0           1           0  
3                 0           0               0           1           0  
4                 0           0               0           1           0  

[5 rows x 98 columns]

In [106]:
X2 = encoded_data.iloc[:,1:99]
y2 = encoded_data.iloc[:,0]

X2 = np.array(X2).reshape((-1,97))
y2 = np.array(y2).reshape((-1,1))

In [107]:
s = []
for i in range(1,98):
  pca2 = PCA(n_components=i,random_state=0)
  pca2.fit(X2)
  x2=pca2.transform(X2)

  model3 = LinearRegression()
  model3.fit(x2,y2)

  pred3 = model3.predict(x2)

  s.append(r2_score(y2, pred3))
  print("With",i,"columns:",s[i-1])

With 1 columns: 0.020463447476670615
With 2 columns: 0.3463746179965892
With 3 columns: 0.354844220694815
With 4 columns: 0.48399829069977074
With 5 columns: 0.5245233822364945
With 6 columns: 0.7513895392062465
With 7 columns: 0.7533272011494492
With 8 columns: 0.8157224608984136
With 9 columns: 0.8285585281684755
With 10 columns: 0.8533899276476711
With 11 columns: 0.8533802490274434
With 12 columns: 0.853538925134046
With 13 columns: 0.8606554599708126
With 14 columns: 0.9009467871524242
With 15 columns: 0.901733068102241
With 16 columns: 0.9029650988621798
With 17 columns: 0.9180193209118185
With 18 columns: 0.9184534308134052
With 19 columns: 0.9214494192319863
With 20 columns: 0.9226204802718533
With 21 columns: 0.9221552964316928
With 22 columns: 0.9236437276500394
With 23 columns: 0.9234788577343084
With 24 columns: 0.9253637402029359
With 25 columns: 0.9256564901542834
With 26 columns: 0.9304535534119325
With 27 columns: 0.9306495777117311
With 28 columns: 0.9336821836403763
W

In [132]:
for i in s:
  if i>=0.992:
      print("With",i,"we have:",s.index(i)+1,"columns out of 98")

With 0.992176471462652 we have: 71 columns out of 98
With 0.9921857928152665 we have: 72 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we have: 73 columns out of 98
With 0.9924082614213805 we ha

# **CONCLUSION**

### **So we are able to achieve 99.24% of the prediction of the price with 73 columns out of 98**